### PR TITLE
Factor out iterator over options dictionaries

### PR DESCRIPTION
**Pull request by Ericson2314**
Otherwise it's easy to forget one. In fact we did forget one: `backend_options` in `validate_option_value`.

### CODE REVIEW

File: mesonbuild/coredata.py [hunk 1]
```diff
@@ -476,8 +477,14 @@ def set_buildtype_from_others(self):
             mode = 'custom'
         self.builtins['buildtype'].set_value(mode)
 
+    def _get_all_nonbuiltin_options(self):
+        yield self.backend_options;
+        yield self.user_options;
```
Comment (ignatenkobrain) [inline mesonbuild/coredata.py]: __[Flake8]__ <sub>posted by Sider</sub>

[E703] statement ends with a semicolon

File: mesonbuild/coredata.py [hunk 2]
```diff
@@ -476,8 +477,14 @@ def set_buildtype_from_others(self):
...
+        yield self.compiler_options;
+        yield self.base_options;
```
Comment (ignatenkobrain) [inline mesonbuild/coredata.py]: __[Flake8]__ <sub>posted by Sider</sub>

[E703] statement ends with a semicolon

File: mesonbuild/coredata.py [hunk 3]
```diff
@@ -476,8 +477,14 @@ def set_buildtype_from_others(self):
             mode = 'custom'
         self.builtins['buildtype'].set_value(mode)
 
+    def _get_all_nonbuiltin_options(self):
+        yield self.backend_options;
+        yield self.user_options;
+        yield self.compiler_options;
```
Comment (ignatenkobrain) [inline mesonbuild/coredata.py]: __[Flake8]__ <sub>posted by Sider</sub>

[E703] statement ends with a semicolon

File: mesonbuild/coredata.py [hunk 4]
```diff
@@ -476,8 +477,14 @@ def set_buildtype_from_others(self):
             mode = 'custom'
         self.builtins['buildtype'].set_value(mode)
 
+    def _get_all_nonbuiltin_options(self):
+        yield self.backend_options;
```
Comment (ignatenkobrain) [inline mesonbuild/coredata.py]: __[Flake8]__ <sub>posted by Sider</sub>

[E703] statement ends with a semicolon

**Comment by Ericson2314** [comment]
Thanks!

**Pull request merged**